internal/orchestrator/commiter.go

```diff
@@ -42,13 +45,203 @@ func (c *Commiter) Start() {
+	var blockNumbers []uint64
+	for i := startBlock; i <= endBlock; i++ {
+		blockNumbers = append(blockNumbers, i)
+	}
+	return blockNumbers, nil
+}
+
+func (c *Commiter) getSequentialBlocksToCommit() ([]common.Block, error) {
+	blocksToCommit, err := c.getBlockNumbersToCommit()
+	if err != nil {
+		return nil, fmt.Errorf("error determining blocks to commit: %v", err)
+	}
+	blocks, err := c.storage.DBStagingStorage.GetBlocks(storage.QueryFilter{BlockNumbers: blocksToCommit})
+	if err != nil {
+		return nil, fmt.Errorf("error fetching blocks to commit: %v", err)
+	}
+	if len(blocks) == 0 {
+		return nil, nil
+	}
+
+	// Sort blocks by block number
+	sort.Slice(blocks, func(i, j int) bool {
+		return blocks[i].Number < blocks[j].Number
+	})
+
+	var sequentialBlocks []common.Block
+	expectedBlockNumber := blocks[0].Number
+
+	for _, block := range blocks {
+		if block.Number != expectedBlockNumber {
+			// Gap detected, stop here
+			break
+		}
+		sequentialBlocks = append(sequentialBlocks, block)
+		expectedBlockNumber++
+	}
+
+	return sequentialBlocks, nil
+}
+
+func (c *Commiter) commit(blocks []common.Block) error {
+	blockNumbers := make([]uint64, len(blocks))
+	for i, block := range blocks {
+		blockNumbers[i] = block.Number
+	}
+
+	logs, transactions, err := c.getStagingDataForBlocks(blockNumbers)
+	if err != nil {
+		return fmt.Errorf("error fetching staging data: %v", err)
+	}
+
+	if err := c.saveDataToMainStorage(blocks, logs, transactions); err != nil {
+		return fmt.Errorf("error saving data to main storage: %v", err)
+	}
+
+	if err := c.deleteDataFromStagingStorage(blocks, logs, transactions); err != nil {
+		return fmt.Errorf("error deleting data from staging storage: %v", err)
+	}
+
+	return nil
+}
+
+func (c *Commiter) getStagingDataForBlocks(blockNumbers []uint64) (logs []common.Log, transactions []common.Transaction, err error) {
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	var logErr, txErr error
+
+	go func() {
+		defer wg.Done()
+		logs, logErr = c.storage.DBStagingStorage.GetEvents(storage.QueryFilter{BlockNumbers: blockNumbers})
+	}()
+
+	go func() {
+		defer wg.Done()
+		transactions, txErr = c.storage.DBStagingStorage.GetTransactions(storage.QueryFilter{BlockNumbers: blockNumbers})
+	}()
+
+	wg.Wait()
+
+	if logErr != nil {
+		return nil, nil, fmt.Errorf("error fetching logs: %v", logErr)
+	}
+	if txErr != nil {
+		return nil, nil, fmt.Errorf("error fetching transactions: %v", txErr)
+	}
+
+	return logs, transactions, nil
+}
+
+func (c *Commiter) saveDataToMainStorage(blocks []common.Block, logs []common.Log, transactions []common.Transaction) error {
+	var commitWg sync.WaitGroup
+	commitWg.Add(3)
+
+	var commitErr error
+	var commitErrMutex sync.Mutex
+
+	go func() {
+		defer commitWg.Done()
+		if err := c.storage.DBMainStorage.InsertBlocks(blocks); err != nil {
+			commitErrMutex.Lock()
+			commitErr = fmt.Errorf("error inserting blocks: %v", err)
+			commitErrMutex.Unlock()
```

Comment:
shouldn't the mutex include the insert operation?

Reply:
This would make the inserts sequantial, not parallel